src/CommunityToolkit/Xamarin.CommunityToolkit/Views/SegmentedView/SegmentedView.shared.cs

```diff
@@ -0,0 +1,262 @@
+
+		string GetDisplayMember(object item)
+		{
+			if (ItemDisplayBinding == null)
+				return item.ToString();
+
+			return (string)GetValue(DisplayProperty);
+		}
+
+		static void OnItemsSourceChanged(BindableObject bindable, object oldValue, object newValue)
+		{
+			((SegmentedView)bindable)?.OnItemsSourceChanged((IEnumerable)oldValue, (IEnumerable)newValue);
+		}
+
+		void OnItemsSourceChanged(IEnumerable oldValue, IEnumerable newValue)
+		{
+			if (oldValue is INotifyCollectionChanged oldObservable)
+				oldObservable.CollectionChanged -= CollectionChanged;
+
+			if (newValue is INotifyCollectionChanged newObservable)
+				newObservable.CollectionChanged += CollectionChanged;
+
+			if (newValue != null)
+				ResetItems();
+		}
+
+		void CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
+		{
+			switch (e.Action)
+			{
+				case NotifyCollectionChangedAction.Add:
+					AddItems(e);
+					break;
+				case NotifyCollectionChangedAction.Remove:
+					RemoveItems(e);
+					break;
+				default:
+					ResetItems();
+					break;
+			}
+		}
+
+		void AddItems(NotifyCollectionChangedEventArgs e)
+		{
+			var index = e.NewStartingIndex < 0 ? Items.Count() : e.NewStartingIndex;
+			foreach (var newItem in e.NewItems)
+				((LockableObservableListWrapper)Items).InternalInsert(index++, GetDisplayMember(newItem));
+		}
+
+		void RemoveItems(NotifyCollectionChangedEventArgs e)
+		{
+            var itemsCount = Items.Count();
+
+			var index = e.OldStartingIndex < itemsCount ? e.OldStartingIndex : itemsCount;
+
+			foreach (var _ in e.OldItems)
+				((LockableObservableListWrapper)Items).InternalRemoveAt(index--);
+		}
+
+		void ResetItems()
+		{
+			if (ItemsSource == null)
+				return;
+
+			((LockableObservableListWrapper)Items).InternalClear();
+
+			foreach (var item in ItemsSource)
+				((LockableObservableListWrapper)Items).InternalAdd(GetDisplayMember(item));
+
+			UpdateSelectedItem(SelectedIndex);
+		}
+
+		static void OnSelectedItemChanged(BindableObject bindable, object oldValue, object newValue)
+		{
+			var segments = (SegmentedView)bindable;
+			segments.SelectedItem = newValue;
+		}
+
+		void UpdateSelectedItem(int index)
+		{
+			if (index == -1)
+			{
+				SelectedItem = null;
+				return;
+			}
+
+			if (ItemsSource != null)
+			{
+				SelectedItem = ItemsSource.ElementAt(index);
+				return;
+			}
+
+			SelectedItem = Items.ElementAt(index);
+		}
+
+		public int SelectedIndex
+		{
+			get => (int)GetValue(SelectedIndexProperty);
+			set => SetValue(SelectedIndexProperty, value);
+		}
+
+		static void OnSegmentSelected(BindableObject bindable, object oldValue, object newValue)
+		{
+			if (!(bindable is SegmentedView segment))
+				return;
+
+			if (!int.TryParse(newValue?.ToString(), out var index))
+				index = 0;
+
+			segment.OnSelectedIndexChanged(segment, index);
+			segment.SelectedItem = segment?.Items.ElementAt(index);
```

Comment:
An `System.ArgumentOutOfRangeException` exception is thrown here with Index=3 and Children.Count = 3 when switching between Image and Text mode in the sample, also noticed Image count is 6 instead of 3, the double so could be that we are not unsubscribing to an event changed or subscribing more than one time.

Reply:
Hmm... I thought I fixed that. I couldn't reproduce it. Let me try to break it again.

Reply:
Did we reproduce this? Is it fixed? :)